### PR TITLE
feat(encyclopedia): add provenance fields to Beer model (epic #693 part 1/n)

### PR DESCRIPTION
**Pull request by benoit-bremaud**
## Summary

Part **1/n** of [Epic #693 — DB schema + seed for Scan Tranche 2](https://github.com/benoit-bremaud/brasse-bouillon/issues/693). Enriches the \`beers\` table (Python + Alembic) with the four provenance columns the Scan Tranche 2 matching algorithm will read, and that anticipate the v0.2 community contribution flow per **ADR-0001** (shapes anticipate evolution).

Kept small on purpose: only the \`Beer\` model + one migration. Further parts (Recipe quality fields in packages/api, scan_catalog bridge, 501 API stubs, seed scaffolding) ship as separate PRs.

### ``db/models/beer.py`` — 4 new fields

| Column | Type | Nullable | Default | Purpose |
|---|---|---|---|---|
| ``source`` | ``VARCHAR(20)`` | ❌ | ``'internal'`` | Canonical discriminant: \`'openfoodfacts' \| 'internal' \| 'community'\`. CHECK constraint enforces the whitelist at the DB level. |
| ``contributed_by`` | ``UUID`` | ✅ | — | Loose link to the users table in ``packages/api`` (no FK — services may run against separate DBs). NULL everywhere in v0.1. |
| ``contributed_at`` | ``TIMESTAMPTZ`` | ✅ | — | Populated by the v0.2+ community contribution flow. NULL in v0.1. |
| ``approved_at`` | ``TIMESTAMPTZ`` | ✅ | — | Populated by the v0.2+ moderation flow. NULL in v0.1. |

### Migration ``002_beer_provenance_fields.py``

- Adds the 4 columns.
- \`source\` column carries \`server_default='internal'\` so existing rows are **backfilled on the same DDL pass** (no separate data migration).
- Adds \`ck_beers_source_valid\` CHECK constraint duplicating the SQLAlchemy whitelist. Two-layer defence: ORM rejects at insert, DB rejects at raw SQL.

## Design note — coexistence with ``EntitySource``

The polymorphic ``EntitySource`` tracking table already exists in the schema (\`db/models/source.py\`) and records detailed per-import provenance with \`raw_data\` payloads. The new \`Beer.source\` column is **not a replacement** — it is a cheap shorthand that downstream consumers (mobile app, matching algo, \`GET /beers/:barcode\`) branch on **without a join**. Both layers serve different use cases:

- ``Beer.source`` — index-friendly, 1-row-1-read, used by business logic.
- ``EntitySource`` — polymorphic, keeps raw payloads, used for audit + reprocessing.

## Tests

- ✅ ``pytest -q tests/`` — **64/64 passing** (no regression).
- ✅ ``ruff check`` — clean on the modified files.
- ✅ ``python -m py_compile`` — syntax valid on model + migration.
- ⏳ DB-level smoke test (running the migration against PostgreSQL) — requires a live DB and is deferred to CI.

## Checklist

- [x] Happy path + sad path + edge cases covered — 64 existing tests still green; DB-level CHECK constraint covers the sad path (invalid source).
- [x] \`tsc --noEmit\` passes — N/A (Python package).
- [x] Build passes — ruff + py_compile clean.
- [x] Existing tests still green — 64/64.
- [x] No \`any\`, no inline styles introduced — N/A (Python).
- [ ] \`PROJECT_LOG.md\` updated — bundled with the next substantive PR per the log-bundling convention.

## Files

- ``packages/beer-encyclopedia/db/models/beer.py`` — +56 lines (fields + CHECK constraint + detailed docstring).
- ``packages/beer-encyclopedia/migrations/versions/002_beer_provenance_fields.py`` — new, 88 lines.

## Related

- Part of [Epic #693](https://github.com/benoit-bremaud/beer-encyclopedia/issues/693).
- Unblocks [#701 — Seed 6 demo beers](https://github.com/benoit-bremaud/brasse-bouillon/issues/701) (which needs the ``source`` column).
- Prepares [#696 — Backend proxy OpenFoodFacts](https://github.com/benoit-bremaud/brasse-bouillon/issues/696) (which imports beers with \`source='openfoodfacts'\`).
- [ADR-0001 — Build for today, design for tomorrow](docs/architecture/decisions/0001-build-for-today-design-for-tomorrow.md) — clause 2 (shapes anticipate evolution).

### CODE REVIEW

File: packages/beer-encyclopedia/migrations/versions/002_beer_provenance_fields.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    # Add the four new columns. ``source`` ships with a server default
+    # so existing rows are backfilled to 'internal' without requiring a
+    # data migration pass.
+    op.add_column(
+        "beers",
+        sa.Column(
+            "source",
+            sa.String(length=20),
+            nullable=False,
+            server_default="internal",
```
[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/beer-encyclopedia/migrations/versions/002_beer_provenance_fields.py]: Agreed — **Must Have**. `server_default="internal"` was emitted as `DEFAULT internal` (unquoted), which fails on both PostgreSQL and SQLite. Switched to `sa.text("'internal'")` so the DDL is `DEFAULT 'internal'` with quotes, matching the convention from migration 001 (`sa.text("'pending'")`). Fixed in `2e4ba2e`.

File: packages/beer-encyclopedia/db/models/beer.py [hunk 2]
```diff
@@ -45,6 +80,25 @@ class Beer(Base, UUIDMixin, TimestampMixin):
     is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
     is_verified: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
 
+    # Provenance fields (ADR-0001 clause 2: shapes anticipate evolution).
+    source: Mapped[str] = mapped_column(
+        String(20),
+        nullable=False,
+        default="internal",
+        server_default="internal",
+    )
```
[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/beer-encyclopedia/db/models/beer.py:97]: Same point as on the migration — **Must Have**. Applied the same `sa.text("'internal'")` fix on the model so `create_all` / autogenerate comparisons emit the same quoted literal. Fixed in `2e4ba2e`.

File: packages/beer-encyclopedia/db/models/beer.py [hunk 3]
```diff
@@ -19,14 +29,39 @@
     from db.models.tasting_profile import TastingProfile
 
 
+# The three canonical provenance values a Beer row can carry. Kept as a
+# module-level constant so migration and validation logic reference a
+# single source of truth.
+BEER_SOURCE_VALUES = ("openfoodfacts", "internal", "community")
```
[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/beer-encyclopedia/db/models/beer.py:36]: Good catch — **Should Have**. The constant was declared as 'single source of truth' but the CHECK constraint hardcoded the values separately. Now the model's `__table_args__` derives the constraint expression from `BEER_SOURCE_VALUES` via a `format()` call → Python tuple and DDL can never drift. The migration keeps its own copy (documented as intentionally duplicated) so historical migrations stay runnable if the model's tuple evolves — a compromise between DRY and migration independence. Fixed in `2e4ba2e`.

File: packages/beer-encyclopedia/migrations/versions/002_beer_provenance_fields.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    )
+
+    # Enforce the canonical provenance values at the DB level. Duplicates
+    # the whitelist on the SQLAlchemy model so both layers reject invalid
+    # inserts — the CHECK is the last line of defence against direct SQL.
+    op.create_check_constraint(
+        "ck_beers_source_valid",
+        "beers",
+        "source IN ('openfoodfacts', 'internal', 'community')",
+    )
```
[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/beer-encyclopedia/migrations/versions/002_beer_provenance_fields.py]: Agreed — **Should Have**. The original claim was misleading: `Mapped[str]` only type-checks the Python type, not the enum membership, so an ORM assignment like `beer.source = 'typo'` would travel silently to flush time and only fail as an `IntegrityError`. Added a `@validates('source')` method on `Beer` that raises `ValueError` at assignment time when a value is outside `BEER_SOURCE_VALUES`, with a readable message. Now both layers genuinely reject: ORM at assignment time (clear error at the call site) + DB at flush time (last-line-of-defence against direct SQL). Fixed in `2e4ba2e`.

[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/beer-encyclopedia/migrations/versions/002_beer_provenance_fields.py]: Valid catch — **Must Have (P2)**. You're right that `op.create_check_constraint` on an existing table is not portable to SQLite, and migration 001 explicitly handles the PG/SQLite split via `_is_postgres()`. Wrapped the whole upgrade (add_column x4 + create_check_constraint) in `op.batch_alter_table('beers')`, which transparently rewrites the table on SQLite under the hood while staying a no-op layer on PostgreSQL. Applied the same wrapper on `downgrade` for symmetry. Fixed in `2e4ba2e`.

**Comment by benoit-bremaud** [comment]
### Review wrap-up — ready to merge

**CI**: 5/5 green (`changes`, `security-audit`, `beer-encyclopedia` _(ruff + pytest 64/64 + coverage)_, `SonarCloud Scan`, `SonarCloud Code Analysis`) + 3 expected skips (mobile-app, api, website). 0 failure.

**Reviews — 1 round (Copilot + Codex)**:
- **Copilot** — 4 comments (2 Must Have + 2 Should Have):
  1. **Must Have** — unquoted `server_default="internal"` in the migration emits `DEFAULT internal` which is invalid. Fixed with `sa.text("'internal'")`.
  2. **Must Have** — same bug on the model; fixed identically.
  3. **Should Have** — `BEER_SOURCE_VALUES` was declared as a source of truth but hardcoded separately in the CHECK constraint. Model now derives the constraint expression via `format()` on the tuple. Migration keeps its own intentional copy (documented) so historical migrations stay runnable if the model evolves.
  4. **Should Have** — "both layers reject" claim was inaccurate. Added `@validates('source')` on the model for real ORM-layer rejection at assignment time, complementing the DB CHECK.
- **Codex** — 1 **P2 Must Have**: `op.create_check_constraint` on an existing table is not portable to SQLite (migrations run on SQLite too per `_is_postgres()` in 001). Fixed with `op.batch_alter_table('beers')` which transparently rewrites the table under SQLite.
- All addressed in `2e4ba2e`.

**Summary**: 3 Must Have + 2 Should Have → all addressed, 0 Disagree, 0 open.

**Ready to merge** (squash + delete branch).

After merge: 4 sub-parts of Epic [#693](https://github.com/benoit-bremaud/brasse-bouillon/issues/693) remain (Recipe quality fields, scan_catalog bridge, 501 stubs, seed skeleton) — each shipping as a separate small PR to keep reviews light.
